iceoryx_examples/automotive_soa/include/minimal_proxy.hpp

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_AUTOMOTIVE_SOA_MINIMAL_PROXY_HPP
+#define IOX_EXAMPLES_AUTOMOTIVE_SOA_MINIMAL_PROXY_HPP
+
+#include "topic.hpp"
+
+#ifdef USE_UDS
+#include "owl/kom/event_subscriber_uds.hpp"
+#else
+#include "owl/kom/event_subscriber.hpp"
+#endif
+#include "owl/kom/field_subscriber.hpp"
+#include "owl/kom/method_client.hpp"
+#include "owl/runtime.hpp"
+#include "owl/types.hpp"
+
+class MinimalProxy
+{
+  public:
+    static constexpr char m_serviceIdentifier[] = "MinimalSkeleton";
+
+    MinimalProxy(owl::kom::ProxyHandleType& handle) noexcept;
+    ~MinimalProxy() noexcept = default;
+
+    MinimalProxy(const MinimalProxy&) = delete;
+    MinimalProxy(MinimalProxy&&) = delete;
+    MinimalProxy& operator=(const MinimalProxy&) = delete;
+    MinimalProxy& operator=(MinimalProxy&&) = delete;
+
+    static owl::kom::FindServiceHandle StartFindService(owl::kom::FindServiceHandler<owl::kom::ProxyHandleType> handler,
+                                                        owl::kom::InstanceIdentifier& instanceIdentifier) noexcept;
+
+    static void StopFindService(owl::kom::FindServiceHandle handle) noexcept;
+
+    static owl::kom::ServiceHandleContainer<owl::kom::ProxyHandleType>
+    FindService(owl::kom::InstanceIdentifier& instanceIdentifier) noexcept;
+
+    const owl::core::String m_instanceIdentifier;
+#ifdef USE_UDS
```

Comment:
Why not use a variant here? Or implement this via inheritance? Or abstract this directly in the class? The macro is in my opinion not a clean approach here.

When we want to define this at compile-time we could also define a global value which is used as template parameter in `EventSubscriber`.
For instance we state in cmake `target_compile_definitions(my_target MACRO_IPC_TYPE=Transmission::UDS)`

And in C++ we write
```cxx
enum class Transmission {
  UDS
  WHATEVER
};

template<typename T, Transmission TransmissionType>
class EventSubscriber;

template<typename T>
class EventSubscriber<T, TransmissionType::UDS> { /* implementation */ };

template<typename T>
class EventSubscriber<T, TransmissionType::WHATEVER> { /* implementation */ };
```

This could then be used here like
```cxx
class MinimalProxy {
  static constexpr Transmission PROXY_IPC = MACRO_IPC_TYPE;
  owl::kom::EventSubscriber<TimestampTopic1Byte, PROXY_IPC> m_event{....
}
```

(I admit the names could be better ;) )

Reply:
Yeah, you're right. A compile-time polymorphism solution is nicer. I implemented what you described, but with different names ;)